api/kyverno/v1/spec_types.go

```diff
@@ -30,8 +31,9 @@ func (a ValidationFailureAction) Audit() bool {
 
 type ValidationFailureActionOverride struct {
 	// +kubebuilder:validation:Enum=audit;enforce
-	Action     ValidationFailureAction `json:"action,omitempty" yaml:"action,omitempty"`
-	Namespaces []string                `json:"namespaces,omitempty" yaml:"namespaces,omitempty"`
+	Action            ValidationFailureAction `json:"action,omitempty" yaml:"action,omitempty"`
+	Namespaces        []string                `json:"namespaces,omitempty" yaml:"namespaces,omitempty"`
+	NamespaceSelector *metav1.LabelSelector   `json:"namespaceSelector,omitempty" yaml:"namespaceSelector,omitempty"`
```

Comment:
@eddycharly @chipzoller - how do we add new attributes to the policy definition? Are new fields go into v2beta1 only or we need to update both versions?

Reply:
We should try and ensure that any new fields go into v2beta1 only at this point.

Reply:
Does that mean such `namespaceSelector` feature won't bring up in `v1`?
And so ideally if we want to introduce this feature later, it is recommended to upgrade our policies to `v2beta1` ?

Reply:
I can migrate this feature to `v2beta1` only. Let me know if we wish so.

Reply:
Agreed with Chip, any new attributes should go into v2 version. `v2beta1` is the first beta version in v2 which has pretty much the same fields as v1 (v2 removed deprecated fields). Now to add new fields, should we bump to `v2beta2`?

Per [Kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-parts-of-the-api):

>Rule #2: API objects must be able to round-trip between API versions in a given release without information loss, with the exception of whole REST resources that do not exist in some versions.
>
>For example, an object can be written as v1 and then read back as v2 and converted to v1, and the resulting v1 resource will be identical to the original. The representation in v2 might be different from v1, but the system knows how to convert between them in both directions. **Additionally, any new field added in v2 must be able to round-trip to v1 and back, which means v1 might have to add an equivalent field or represent it as an annotation.**

Seems like we also need to add any new field to v1 policy either as an equivalent field or an annotation.

So to conclude, new changes need to be added in both v1 and v2beta**X**.

Reply:
@realshuting Sounds good!
So wondering how we can add spec to `v2beta1`? OR do we need to add additional definition in the package of `v2beta1`?
See that the package is currently importing the definition from v1: https://github.com/kyverno/kyverno/blob/v1.8.1/api/kyverno/v2beta1/spec_types.go#L42

Reply:
We can just update both versions of CRDs.